revvy/robot_config.py

```diff
@@ -114,44 +165,14 @@ def from_string(config_string):
         initial_state = dict_get_first(json_config, ['initialState', 'initialstate'])
         if initial_state:
             config.background_initial_state = initial_state
+        else:
+            # If 'initialState' is not set, default is 'running' which will start all
+            # background scripts at the start of play session.
+            config.background_initial_state = 'running'
 
         try:
-            i = 0
-            for script in blockly_list:
-                _log(f'Processing script #{i}')
-                runnable = RobotConfig.create_runnable(script, i)
-
-                assignments = script['assignments']
-                # script names are mostly relevant for logging
-                for analog_assignment in assignments.setdefault('analog', []):
-                    channels = ', '.join(map(str, analog_assignment['channels']))
-                    script_name = f'[script {i}] analog channels {channels}'
-                    priority = analog_assignment['priority']
-                    config.controller.analog.append({
-                        'channels': analog_assignment['channels'],
-                        'script': ScriptDescriptor(script_name, runnable, priority)})
-                    i += 1
-
-                for variable_assignments in assignments.setdefault('variableSlots', []):
-                    variable_slot = variable_assignments['slot']
-                    variable_name = variable_assignments['variable']
-                    config.controller.variable_slots.append({'slot': variable_slot,
-                                                             'variable': variable_name,
-                                                             'script': i,
-                                                             })
-
-                for button_assignment in assignments.setdefault('buttons', []):
-                    button_id = button_assignment['id']
-                    script_name = f'[script {i}] button {button_id}'
-                    priority = button_assignment['priority']
-                    config.controller.buttons[button_id] = ScriptDescriptor(script_name, runnable, priority)
-                    i += 1
-
-                if 'background' in assignments:
-                    script_name = f'[script {i}] background'
-                    priority = assignments['background']
-                    config.background_scripts.append(ScriptDescriptor(script_name, runnable, priority))
-                    i += 1
+            for script_idx, script in enumerate(blockly_list):
+                RobotConfig.process_script(config, script, script_idx)
```

Comment:
I'm just placing a note here that this refactor does change the logic in a minor way:
 - previously, each "button script", "analog script", etc had their own unique index
 - now, each of these scripts have the same script index

If I can believe my own comment saying this is mostly for logging, this is probably fine, but only because the script names aren't important.

Oh, and be aware that some tests will fail now because I asserted on the script names, e.g. [here](https://github.com/RevolutionRobotics/RevvyFramework/blob/master/tests/unit_tests/test_robot_config.py#L63)

Reply:
Having script index incremented differently and separately from script iteration is a readability problem,
I have spent more than required to understand if "i" is important and what it does. Then I figured out this is only name generation, which is mainly for log / debug. This complexity definitely had to be removed, because this parsing script is too complex in itself to also have additional complexity for debug purpose names. Added complexity adds maintainence effort and risk of new bugs, so in general if it is not required - we don't want it